[DependencyInjection] Enable deprecating parameters

HeahDude opened the pull request:
| Q             | A
| ------------- | ---
| Branch?       | 6.3
| Bug fix?      | no
| New feature?  | yes
| Deprecations? | no
| Tickets       | ~
| License       | MIT
| Doc PR        | TODO

Following #47680, this PR allows deprecating parameters either to move to build parameters, or to rename them or just to be able to remove them in a BC way.

```php
// with ContainerBuilder:
$container->setParameter('old_param', 'xxx');
$container->setParameter('new_param', 'yyy');
$container->deprecateParameter(
    'old_param',
    'symfony/xxx',
    '6.2',
    # Fourth arg is optional, default: 'The parameter "%s" is deprecated.'
    'The param "%s" is deprecated use "%new_param%" instead.'
);
```
_____
TODO:
- [x] Waiting for @nicolas-grekas (and others) to challenge the implementation
- [x] Update CHANGELOG
- [x] Add tests

HeahDude committed:
[DependencyInjection] Enable deprecating parameters

stof commented on src/Symfony/Component/DependencyInjection/Dumper/PhpDumper.php:
```diff
@@ -1375,6 +1377,23 @@ public function getRemovedIds(): array
+    {
+        $deprecated = $this->container->getParameterBag()->allDeprecated();
+        if (!$deprecated) {
+            return '';
```
should we really replace the definition of the empty array with nothing in such case ?

HeahDude replied:
Why would we compute a private property if there is nothing to export and assign?

stof commented on src/Symfony/Component/DependencyInjection/Dumper/PhpDumper.php:
```diff
@@ -1536,6 +1555,7 @@ private function addDefaultParametersMethod(): string
 
     public function getParameter(string $name): array|bool|string|int|float|\UnitEnum|null
     {
+        __trigger_deprecation__
```
This is not consistent with the ParameterBag logic, where a parameter is first checked for existence and only then for deprecation

HeahDude replied:
Fixed, thanks!

stof commented on src/Symfony/Component/DependencyInjection/Dumper/PhpDumper.php:
```diff
@@ -1574,13 +1594,23 @@ public function getParameterBag(): ParameterBagInterface
                 $parameters[$name] = $value;
             }
-            $this->parameterBag = new FrozenParameterBag($parameters);
+            $this->parameterBag = new FrozenParameterBag($parameters, $this->deprecatedParameters ?? []);
```
why `??[]` ? We should always have an array IMO.

HeahDude replied:
Because as you noticed in your previous comment the property may not be dumped at all.

nicolas-grekas replied:
we know at dump-time whether the property is going to exist, so let's remove this when not

HeahDude replied:
Fixed.

stof reviewed (changes requested):
This misses the update of the ContainerBuilder merging to copy the deprecations.

HeahDude commented:
Status: needs review

nicolas-grekas commented on src/Symfony/Component/DependencyInjection/Container.php:
```diff
@@ -123,6 +123,14 @@ public function setParameter(string $name, array|bool|string|int|float|\UnitEnum
+    /**
+     * @throws ParameterNotFoundException if the parameter is not defined
+     */
+    public function deprecateParameter(string $name, string $package, string $version, string $message = 'The parameter "%s" is deprecated.')
```
I'd suggest removing this method, the one on the parameter-bag is enough to me

HeahDude replied:
After a quick discussion, we decided to keep it in favor of adding `deprecate()` to the `ParameterBagInterface`.

nicolas-grekas commented on src/Symfony/Component/DependencyInjection/Dumper/PhpDumper.php:
```diff
@@ -1375,6 +1377,23 @@ public function getRemovedIds(): array
+    private function addDeprecatedParameters(): string
+    {
+        $deprecated = $this->container->getParameterBag()->allDeprecated();
+        if (!$deprecated) {
```
`if (!$deprecated = $this->container->getParameterBag()->allDeprecated()) {`

HeahDude replied:
Done.

nicolas-grekas commented on src/Symfony/Component/DependencyInjection/Dumper/PhpDumper.php:
```diff
@@ -1375,6 +1377,23 @@ public function getRemovedIds(): array
+        $code = '';
+        ksort($deprecated);
+        foreach ($deprecated as $param => $deprecation) {
+            $code .= '        '.$this->doExport($param)." => ".$this->doExport($deprecation).",\n";
```
`' => '`

HeahDude replied:
I've changed the export altogether.

nicolas-grekas commented on src/Symfony/Component/DependencyInjection/Dumper/PhpDumper.php:
```diff
@@ -1574,13 +1595,23 @@ public function getParameterBag(): ParameterBagInterface
+$1    $deprecated = $this->deprecatedParameters[$name];
+$1    trigger_deprecation($deprecated['package'], $deprecated['version'], sprintf($deprecated['message'], $name));
+$1}
+EOF, $code);
```
please indent the nowdoc block

HeahDude replied:
Done.

nicolas-grekas commented on src/Symfony/Component/DependencyInjection/Dumper/PhpDumper.php:
```diff
@@ -1574,13 +1595,23 @@ public function getParameterBag(): ParameterBagInterface
+            $code = preg_replace("/( +)__trigger_deprecation__/m", <<<'EOF'
+$1if (isset($this->deprecatedParameters[$name])) {
+$1    $deprecated = $this->deprecatedParameters[$name];
+$1    trigger_deprecation($deprecated['package'], $deprecated['version'], sprintf($deprecated['message'], $name));
```
no need for the sprintf around, just pass the message and the name to trigger_deprecation

HeahDude replied:
Done (in ParameterBag as well).

nicolas-grekas commented on src/Symfony/Component/DependencyInjection/Loader/XmlFileLoader.php:
```diff
@@ -80,6 +80,9 @@ private function loadXml(\DOMDocument $xml, string $path, \DOMNode $root = null)
         $this->parseParameters($xml, $path, $root);
 
+        // deprecated parameters
+        $this->parseDeprecatedParameters($xml, $root);
```
I think deprecated parameters should be declared in DI extensions only. As such, I'd suggest removing all the code related to loaders/configurators

HeahDude replied:
Done and description of the PR updated.

nicolas-grekas commented on src/Symfony/Component/DependencyInjection/ParameterBag/FrozenParameterBag.php:
```diff
@@ -28,8 +28,10 @@ class FrozenParameterBag extends ParameterBag
-    {
+    public function __construct(
+        array $parameters = [],
+        protected array $deprecatedParameters = [],
```
private would be fine
let's remove the `@param` annotation while at it

HeahDude replied:
Done.

nicolas-grekas commented on src/Symfony/Component/DependencyInjection/ParameterBag/ParameterBag.php:
```diff
@@ -81,6 +87,12 @@ public function get(string $name): array|bool|string|int|float|\UnitEnum|null
+        if (isset($this->deprecatedParameters[$name])) {
+            $deprecated = $this->deprecatedParameters[$name];
+
+            trigger_deprecation($deprecated['package'], $deprecated['version'], sprintf($deprecated['message'], $name));
```
no sprintf

nicolas-grekas commented on src/Symfony/Component/DependencyInjection/ParameterBag/ParameterBagInterface.php:
```diff
@@ -18,6 +18,9 @@
  * @author Fabien Potencier <fabien@symfony.com>
+ *
+ * @method array allDeprecated() Returns all deprecated parameters. Not implementing it is deprecated since Symfony 6.2.
+ * @method deprecate(string $name, string $package, string $version, string $message) Deprecates a parameter. Not implementing it is deprecated since Symfony 6.2.
```
do we really need to make this part of the interface? I'd prefer if it could stay on ParameterBag

HeahDude replied:
`ContainerBuilder` exposes now the only public API.

HeahDude commented:
The new implementation does not trigger deprecations in the API itself, so the label can be removed.
Thanks for the reviews.

nicolas-grekas reviewed (approved):
I just squashed+added a commit with some tweaks.
LGTM thanks!

stof commented on src/Symfony/Component/DependencyInjection/ParameterBag/ParameterBag.php:
```diff
@@ -47,6 +48,11 @@ public function all(): array
         return $this->parameters;
     }
 
+    public function allDeprecated(): array
```
should this be marked as `@internal` if you don't consider it part of the public API ? For now, it is public.

nicolas-grekas replied:
I think it's fine having it public.

HeahDude replied:
Maybe we should either mark the `ParameterBag` new methods as internal indeed or document them in the changelog if they are public?

chalasr replied:
Fine as-is to me, no need to talk about it

HeahDude replied:
Since I now use protected (https://github.com/symfony/symfony/pull/47719#issuecomment-1356260382 and https://github.com/symfony/symfony/pull/47719/commits/63d39b0ac7e93ffb5d74e26e53652035cc49c634), maybe we should at least mark the methods as final?
Btw the `FrozenParameterBag` class should also be marked `@final` as of 6.3, wdyt?

nicolas-grekas replied:
I'm fine as is personally. Making it final won't provide us anything.

HeahDude commented on src/Symfony/Component/DependencyInjection/Dumper/PhpDumper.php:
```diff
@@ -1550,6 +1569,10 @@ private function addDefaultParametersMethod(): string
 
     public function getParameter(string $name): array|bool|string|int|float|\UnitEnum|null
     {
+        if (isset(self::DEPRECATED_PARAMETERS[$name])) {
```
@nicolas-grekas this change is a regression regarding @stof's comment that I had addressed.
It is not consistent anymore with the parameter bag logic, where we first check for a parameter existence before triggering a deprecation (https://github.com/symfony/symfony/pull/47719#discussion_r983595581).

HeahDude replied:
It makes me think about removing parameters, when it happens we should remove any related deprecation as well to prevent such case in the dumped container where we would trigger a deprecation before throwing a `ParameterNotFoundException`.
WDYT?

nicolas-grekas replied:
indeed, we should only list params that exist in this array

HeahDude replied:
OK, I've pushed https://github.com/symfony/symfony/pull/47719/commits/81e63a734e787760f62e1c90c26a85207eaec791.

HeahDude commented:
While trying the feature I realized that the `FrozenParameterBag` fails to trigger a deprecation due to property visibility in `parent::get()`.
I've added a test that covers this and fixed it by using `protected`, the property remains immutable anyway thanks to the exception in `deprecate()` (now covered as well).

nicolas-grekas reviewed (commented):
Still :+1: on my side after the last changes
/cc @symfony/mergers any other comments?

chalasr commented:
Thank you @HeahDude.

chalasr commented:
gh went wrong due to some race condition, merged in https://github.com/symfony/symfony/commit/be5fbced648bccd0f7a907c1b980e5fdb89211c0

OskarStark commented:
Looks like there was no doc issue opened 🤔
